output/cloud/expv2/hdr.go

```diff
@@ -104,92 +94,72 @@ func (h *histogram) addToBucket(v float64) {
+		h.Buckets = make(map[uint32]uint32)
 	}
+	h.Buckets[ix]++
 }
 
-func (h *histogram) init(index uint32) {
-	h.FirstNotZeroBucket = index
-	h.LastNotZeroBucket = index
-	h.Buckets = make([]uint32, 1, 32)
-	h.Buckets[0] = 1
-}
+// trackBucket stores the unique seen buckets.
+func (h *histogram) trackBucket(index uint32) {
+	i := sort.Search(len(h.Indexes), func(i int) bool {
+		return h.Indexes[i] > index
+	})
 
-// prependBuckets expands the buckets slice with zeros up to the required index,
-// then it increments the required bucket.
-func (h *histogram) prependBuckets(index uint32) {
-	if h.FirstNotZeroBucket <= index {
-		panic("buckets is already contains the requested index")
+	// insert at the end
+	if len(h.Indexes) == i {
+		h.Indexes = append(h.Indexes, index)
+		return
 	}
 
-	newLen := (h.FirstNotZeroBucket - index) + uint32(len(h.Buckets))
-
-	// TODO: we may consider to swap by sub-groups
-	// e.g  [4, 1] => [4, 1, 0, 0] => [0, 0, 4, 1]
-	// It requires a benchmark if it is better than just copy it.
-
-	newBuckets := make([]uint32, newLen)
-	copy(newBuckets[h.FirstNotZeroBucket-index:], h.Buckets)
-	h.Buckets = newBuckets
-
-	// Update the stats
-	h.Buckets[0] = 1
-	h.FirstNotZeroBucket = index
+	// insert in the middle
+	h.Indexes = append(h.Indexes, 0)     // expand the slice
+	copy(h.Indexes[i+1:], h.Indexes[i:]) // make the space
+	h.Indexes[i] = index                 // set the index
 }
 
-// appendBuckets expands the buckets slice with zeros buckets till the required index,
-// then it increments the required bucket.
-// If the slice has enough capacity then it reuses it without allocate.
-func (h *histogram) appendBuckets(index uint32) {
-	if h.LastNotZeroBucket >= index {
-		panic("buckets is already bigger than requested index")
+// histogramAsProto converts the histogram into the equivalent Protobuf version.
+func histogramAsProto(h *histogram, time int64) *pbcloud.TrendHdrValue {
+	var (
+		counters []uint32
+		spans    []*pbcloud.BucketSpan
+	)
+
+	// allocate only if at least one item is available
+	if len(h.Indexes) > 0 {
+		// init the counters
+		counters = make([]uint32, 1, len(h.Indexes))
+		counters[0] = h.Buckets[h.Indexes[0]]
+		// open the first span
+		spans = append(spans, &pbcloud.BucketSpan{Offset: h.Indexes[0], Length: 1})
 	}
 
-	newLen := index - h.FirstNotZeroBucket + 1
+	for i := 1; i < len(h.Buckets); i++ {
+		counters = append(counters, h.Buckets[h.Indexes[i]])
 
-	if uint32(cap(h.Buckets)) > newLen {
-		// See https://go.dev/ref/spec#Slice_expressions
-		// "For slices, the upper index bound is
-		// the slice capacity cap(a) rather than the length"
-		h.Buckets = h.Buckets[:newLen]
-	} else {
-		newBuckets := make([]uint32, newLen)
-		copy(newBuckets, h.Buckets)
-		h.Buckets = newBuckets
-	}
+		// if the current and the previous indexes are not consecutive
+		// consider as closed the current on-going span and start a new one.
+		if diff := h.Indexes[i] - h.Indexes[i-1]; diff > 1 {
```

Comment:
For more optimal space use this likely should be 3+ as adding 1 span is at least 2 counters (offset and length) .

So we can add at least 2 zeros before we became even. "at least" here is as I don't know if there is some more overhead for this in the protobuf protocol